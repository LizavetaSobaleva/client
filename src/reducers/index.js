import { applyMiddleware, combineReducers } from "redux";
import { configureStore} from "@reduxjs/toolkit"
import { thunk } from "redux-thunk";
import userReducer from "./userReducer";
import fileReducer from "./fileReducer";
import uploadReducer from "./uploadReducer";
import appReducer from "./appReducer";
import { composeWithDevTools } from '@redux-devtools/extension';


const rootReducer = combineReducers({
    user: userReducer,
    files: fileReducer,
    upload: uploadReducer,
    app: appReducer
})

export const store = configureStore({reducer: rootReducer, devTools: composeWithDevTools(applyMiddleware(thunk))})