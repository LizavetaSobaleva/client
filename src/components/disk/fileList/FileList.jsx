import React, { useState } from "react"
import "./fileList.less"
import { useDispatch, useSelector } from "react-redux"
import File from "./file/File"
import { popFromStack, setCurrentDir } from "../../../reducers/fileReducer"
import backIcon from "../../../assets/img/back.png"
import { TransitionGroup, CSSTransition } from "react-transition-group"
import Loader from "../../UI/loader/Loader"


const FileList = () => {
  const dispatch = useDispatch()
  const currentDir = useSelector((state) => state.files.currentDir)
  const loader = useSelector(state => state.app.loader)
  const fileView = useSelector(state => state.app.view)
  const [fieldName, setFieldName] = useState('type');

  const files = useSelector((state) => state.files.files)
  const sortedFiles = files.slice().sort((a,b) => a[fieldName].localeCompare(b[fieldName]))

  const dirStack = useSelector((state) => state.files.dirStack)

  function backClickHandler() {
    dispatch(popFromStack(currentDir));
    const backDirId = dirStack.pop();
    dispatch(setCurrentDir(backDirId));
  }


  if (fileView === 'list') {
    return (
      <>
        <div className="filelist">
          <div className="filelist__header">
            <div className="filelist__type" onClick={() => setFieldName('type')}>Type</div>
            <div className="filelist__name" onClick={() => setFieldName('name')}>Name</div>
            <div className="filelist__date" onClick={() => setFieldName('date')}>Date</div>
            <div className="filelist__size">File size</div>
            <div className="filelist__actions"></div>
          </div>
          
            {currentDir && 
              <div className="filelist__back" onClick={() => backClickHandler()}>
                <img src={backIcon} alt="" className='filelist__img'/>
                Back
              </div>}
  
              {loader && <Loader/> }
            
              <TransitionGroup>
                {sortedFiles.map((file) => 
                  <CSSTransition
                    key={file._id}
                    timeout={500} 
                    classNames={"file"}
                    exit={false}
                    in={!loader}
                    mountOnEnter
                    unmountOnExit
                  >
                    <File file={file} />
                  </CSSTransition>
                )}
              </TransitionGroup>
        </div>
        
        {files.length === 0 &&
          <div className="filelist__empty">
            Files not found
          </div>
        }
      </>
      
    );
  }

  if (fileView === 'plate') {
    return (
      <>
        <div className="fileplate">
          
            {currentDir && 
              <div className="fileplate__back" onClick={() => backClickHandler()}>
                <img src={backIcon} alt="" className='fileplate__img'/>
              </div>}
  
              {loader && <Loader/> }
            
                {sortedFiles.map((file) => 
                  <File file={file} key={file._id} />
                )}
        </div>
        
        {files.length === 0 &&
          <div className="fileplate__empty">
            Files not found
          </div>
        }
      </>
      
    );
  }
  
};

export default FileList;
