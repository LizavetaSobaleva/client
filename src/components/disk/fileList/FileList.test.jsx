import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import FileList from './FileList';
import { setCurrentDir, popFromStack } from '../../../reducers/fileReducer';


jest.mock('./file/File', () => ({ file }) => <div>{file.name}</div>);

const mockStore = configureStore([]);

const files = [
  { _id: '1', name: 'Test File 1', type: 'file', size: 12345, date: '2023-01-01T00:00:00Z' },
  { _id: '2', name: 'Test File 2', type: 'file', size: 12346, date: '2023-01-02T00:00:00Z' },
  { _id: '3', name: 'Test Directory', type: 'dir', size: 0, date: '2023-01-03T00:00:00Z' }
];

let store;

beforeEach(() => {
  store = mockStore({
    files: {
      currentDir: 'root',
      files,
      dirStack: []
    },
    app: {
      view: 'list',
      loader: false
    }
  });
  store.dispatch = jest.fn();
});

const renderComponent = () => render(
  <Provider store={store}>
    <FileList />
  </Provider>
);

test('renders FileList without crashing', () => {
  renderComponent();
  expect(screen.getByText('Test File 1')).toBeInTheDocument();
  expect(screen.getByText('Test File 2')).toBeInTheDocument();
  expect(screen.getByText('Test Directory')).toBeInTheDocument();
});

test('back button works correctly', () => {
  store = mockStore({
    files: {
      currentDir: '2',
      files,
      dirStack: ['root']
    },
    app: {
      view: 'list',
      loader: false
    }
  });
  store.dispatch = jest.fn();
  
  renderComponent();
  const backButton = screen.getByTestId('backBtn');
  fireEvent.click(backButton);
  expect(store.dispatch).toHaveBeenCalledWith(popFromStack('2'));
  expect(store.dispatch).toHaveBeenCalledWith(setCurrentDir('root'));
});

test('loader is displayed correctly', () => {
  store = mockStore({
    files: {
      currentDir: 'root',
      files,
      dirStack: []
    },
    app: {
      view: 'list',
      loader: true
    }
  });
  
  renderComponent();
  expect(screen.getByTestId('loader')).toBeInTheDocument();
});

test('files are sorted by name', () => {
  renderComponent();
  const fileElements = screen.getAllByText(/Test File|Test Directory/);
  expect(fileElements[0]).toHaveTextContent('Test Directory');
  expect(fileElements[1]).toHaveTextContent('Test File 1');
  expect(fileElements[2]).toHaveTextContent('Test File 2');
});

test('empty message is displayed when no files are present', () => {
  store = mockStore({
    files: {
      currentDir: 'root',
      files: [],
      dirStack: []
    },
    app: {
      view: 'list',
      loader: false
    }
  });

  renderComponent();
  expect(screen.getByText('Files not found')).toBeInTheDocument();
});
