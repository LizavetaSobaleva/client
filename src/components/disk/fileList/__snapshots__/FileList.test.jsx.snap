// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`back button works correctly 1`] = `
<DocumentFragment>
  <div
    class="filelist"
  >
    <div
      class="filelist__header"
    >
      <div
        class="filelist__type"
      >
        Type
      </div>
      <div
        class="filelist__name"
      >
        Name
      </div>
      <div
        class="filelist__date"
      >
        Date
      </div>
      <div
        class="filelist__size"
      >
        File size
      </div>
      <div
        class="filelist__actions"
      />
    </div>
    <div
      class="filelist__back"
      data-testid="backBtn"
    >
      <img
        alt=""
        class="filelist__img"
        src="test-file-stub"
      />
      Back
    </div>
    <div>
      <div>
        Test Directory
      </div>
      <div>
        Test File 1
      </div>
      <div>
        Test File 2
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`empty message is displayed when no files are present 1`] = `
<DocumentFragment>
  <div
    class="filelist"
  >
    <div
      class="filelist__header"
    >
      <div
        class="filelist__type"
      >
        Type
      </div>
      <div
        class="filelist__name"
      >
        Name
      </div>
      <div
        class="filelist__date"
      >
        Date
      </div>
      <div
        class="filelist__size"
      >
        File size
      </div>
      <div
        class="filelist__actions"
      />
    </div>
    <div
      class="filelist__back"
      data-testid="backBtn"
    >
      <img
        alt=""
        class="filelist__img"
        src="test-file-stub"
      />
      Back
    </div>
    <div />
  </div>
  <div
    class="filelist__empty"
  >
    Files not found
  </div>
</DocumentFragment>
`;

exports[`files are sorted by name 1`] = `
<DocumentFragment>
  <div
    class="filelist"
  >
    <div
      class="filelist__header"
    >
      <div
        class="filelist__type"
      >
        Type
      </div>
      <div
        class="filelist__name"
      >
        Name
      </div>
      <div
        class="filelist__date"
      >
        Date
      </div>
      <div
        class="filelist__size"
      >
        File size
      </div>
      <div
        class="filelist__actions"
      />
    </div>
    <div
      class="filelist__back"
      data-testid="backBtn"
    >
      <img
        alt=""
        class="filelist__img"
        src="test-file-stub"
      />
      Back
    </div>
    <div>
      <div>
        Test Directory
      </div>
      <div>
        Test File 1
      </div>
      <div>
        Test File 2
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`loader is displayed correctly 1`] = `
<DocumentFragment>
  <div
    class="filelist"
  >
    <div
      class="filelist__header"
    >
      <div
        class="filelist__type"
      >
        Type
      </div>
      <div
        class="filelist__name"
      >
        Name
      </div>
      <div
        class="filelist__date"
      >
        Date
      </div>
      <div
        class="filelist__size"
      >
        File size
      </div>
      <div
        class="filelist__actions"
      />
    </div>
    <div
      class="loader"
      data-testid="loader"
    >
      <div
        class="loader__line"
        data-testid="loaderLine"
      />
    </div>
    <div
      class="filelist__back"
      data-testid="backBtn"
    >
      <img
        alt=""
        class="filelist__img"
        src="test-file-stub"
      />
      Back
    </div>
    <div>
      <div>
        Test Directory
      </div>
      <div>
        Test File 1
      </div>
      <div>
        Test File 2
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders FileList without crashing 1`] = `
<DocumentFragment>
  <div
    class="filelist"
  >
    <div
      class="filelist__header"
    >
      <div
        class="filelist__type"
      >
        Type
      </div>
      <div
        class="filelist__name"
      >
        Name
      </div>
      <div
        class="filelist__date"
      >
        Date
      </div>
      <div
        class="filelist__size"
      >
        File size
      </div>
      <div
        class="filelist__actions"
      />
    </div>
    <div
      class="filelist__back"
      data-testid="backBtn"
    >
      <img
        alt=""
        class="filelist__img"
        src="test-file-stub"
      />
      Back
    </div>
    <div>
      <div>
        Test Directory
      </div>
      <div>
        Test File 1
      </div>
      <div>
        Test File 2
      </div>
    </div>
  </div>
</DocumentFragment>
`;
